public/app/plugins/panel/text/TextPanel.tsx

```diff
@@ -1,127 +1,90 @@
-// Libraries
 import { css, cx } from '@emotion/css';
 import DangerouslySetHtmlContent from 'dangerously-set-html-content';
-import { debounce } from 'lodash';
-import React, { PureComponent } from 'react';
+import React, { useMemo } from 'react';
+import { useLocation } from 'react-router-dom';
 
-import { GrafanaTheme2, PanelProps, renderTextPanelMarkdown, textUtil } from '@grafana/data';
-// Utils
-import { CustomScrollbar, CodeEditor, stylesFactory, ThemeContext } from '@grafana/ui';
+import { GrafanaTheme2, PanelProps, renderTextPanelMarkdown, textUtil, InterpolateFunction } from '@grafana/data';
+import { CustomScrollbar, CodeEditor, useStyles2 } from '@grafana/ui';
 import config from 'app/core/config';
 
-// Types
 import { defaultCodeOptions, PanelOptions, TextMode } from './models.gen';
 
 export interface Props extends PanelProps<PanelOptions> {}
 
-interface State {
-  html: string;
-}
-
-export class TextPanel extends PureComponent<Props, State> {
-  declare context: React.ContextType<typeof ThemeContext>;
-  static contextType = ThemeContext;
-
-  constructor(props: Props) {
-    super(props);
-
-    this.state = {
-      html: this.processContent(props.options),
-    };
-  }
-
-  updateHTML = debounce(() => {
-    const html = this.processContent(this.props.options);
-    if (html !== this.state.html) {
-      this.setState({ html });
-    }
-  }, 150);
+export function TextPanel({ options, replaceVariables, width, height }: Props) {
+  const styles = useStyles2(getStyles);
+  const location = useLocation();
```

Comment:
^^ required so we get a callback when variables change

Reply:
@ryantxu it should automatically get re-rendered if a variable that it includes in it's definition change. 

Is that not the case?

Reply:
Ahh -- the issue was that with `useMemo` there are no obvious props to invalidate.  I went back to a debouce approach, and have the invalidate whenever props change (not the options within it)